File: test_parse.py
import re

raw_text = """100.00
Paid to
PRINCIPAL BVRIT NARSAPUR
] Banking name: PRINCIPAL BVRIT NARSAPUR
15 September 2025, 1:07 pm"""

lines = [l.strip() for l in raw_text.split('\n') if l.strip()]
transactions = []

for i, line in enumerate(lines):
    amt_match = re.search(r'(?:[₹$£]|rs\.?|paid|amount|total|-)?\s*(\d+(?:,\d{3})*(?:\.\d{2})?)', line, re.IGNORECASE)
    if amt_match:
        amt_str = amt_match.group(1).replace(',', '')
        amt = float(amt_str)
        
        has_currency_symbol = bool(re.search(r'(?:[₹$£]|rs\.?|paid)', line, re.IGNORECASE))
        if amt > 0 and (has_currency_symbol or '.' in line):
            desc = "Unknown Merchant"
            
            if i > 0 and len(lines[i-1]) > 2 and not bool(re.search(r'\d{3}', lines[i-1])):
                desc = lines[i-1].strip()
            else:
                stripped_line = line.replace(amt_match.group(0), '').strip()
                if len(stripped_line) > 2 and not bool(re.search(r'^[\d:\-\/]+$', stripped_line)):
                    desc = stripped_line
                elif i + 1 < len(lines) and not bool(re.search(r'\d', lines[i+1])):
                    desc = lines[i+1].strip()
            
            desc = re.sub(r'[^a-zA-Z0-9\s&*]', '', desc).strip()
            if not desc:
                desc = 'Unknown Scanned Merchant'
            
            transactions.append({
                'description': desc,
                'amount': amt
            })

print(transactions)

File: test_parse.js
const Helpers = { todayStr: () => '2026-04-11' };

function parseHeuristics(rawText) {
    const lines = rawText.split('\n').map(l => l.trim()).filter(Boolean);
    const transactions = [];
    const today = Helpers.todayStr();

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const amtMatch = line.match(/(?:[₹$£]|rs\.?|paid|amount|total|-)?\s*(\d+(?:,\d{3})*(?:\.\d{2})?)/i);
      
      if (amtMatch) {
        const amtStr = amtMatch[1].replace(/,/g, '');
        const amt = parseFloat(amtStr);

        const hasCurrencySymbol = /(?:[₹$£]|rs\.?|paid)/i.test(line);
        if (amt > 0 && (hasCurrencySymbol || line.includes('.'))) {
          let desc = "Unknown Merchant";
          
          if (i > 0 && lines[i-1].length > 2 && !/\d{3}/.test(lines[i-1])) {
            desc = lines[i-1].trim();
          } else {
            const strippedLine = line.replace(amtMatch[0], '').trim();
            if (strippedLine.length > 2 && !/^[\d:\-\/]+$/.test(strippedLine)) {
               desc = strippedLine;
            } else if (i + 1 < lines.length && !/\d/.test(lines[i+1])) {
               desc = lines[i+1].trim();
            }
          }

          desc = desc.replace(/[^a-zA-Z0-9\s&*]/g, '').slice(0, 50).trim() || 'Unknown Scanned Merchant';

          transactions.push({
            date: today,
            description: desc,
            amount: amt
          });
        }
      }
    }

    const uniqueTxns = [];
    const seen = new Set();
    for (const t of transactions) {
      const key = `${t.amount}-${t.description.toLowerCase()}`;
      if (!seen.has(key)) {
        seen.add(key);
        uniqueTxns.push(t);
      }
    }

    return uniqueTxns.slice(0, 20);
}

const sample = `
Uber ride
₹ 250.00
Date: 11-04-2026
Zomato
Paid 450
Amazon
Amount: $ 1,200.50
`;

console.log(parseHeuristics(sample));
